addons/common/fnc_runQTE.sqf

```diff
@@ -0,0 +1,163 @@
+#include "\a3\ui_f\hpp\defineDIKCodes.inc"
+/*
+	 * Author: john681611
+	 * Runs a Quick time Event.
+	 *
+	 * Arguments:
+	 * 0: Object QTE is attached to <OBJECT>
+	 * 1: Arguments, passed to condition, fail and finish <ARRAY>
+	     * 2: On Display: Code callback on displayable event. <CODE, STRING>
+	 * 2: On Finish: Code called or STRING raised as event. <CODE, STRING>
+	 * 3: On Failure: Code called or STRING raised as event. <CODE, STRING>
+	 * 4: QTE Seqence <ARRAY>
+	 * 5: max interaction distance from attached object <NUMBER> (default: 10) 
+	 * 6: timout <NUMBER> (default: 30) 
+	 *
+	 * Return Value:
+	 * None
+	 *
+	 * Example:
+	 * [car, [], {
+		hint "Finished!"
+	}, {
+		hint "Failure!"
+	}, ["↑", "↓", "→", "←"]] call ace_common_fnc_runQTE
+	 *
+	 * Public: Yes
+ */
+
+params ["_object", "_args", "_onDisplay", "_onFinish", "_onFail", "_qte_seqence", ["_max_distance", 10], ["_timeout", 30]];
+
+if (!GVAR(settingsInitFinished)) exitWith {
+	// only run this after the settings are initialized
+	GVAR(runAtSettingsInitialized) pushBack [FUNC(runQTE), _this];
+};
+
+if (GVAR(QTERunning)) exitWith {
+	TRACE_1("QTE already running qeueing up", GVAR(QTERunning));
+	[{
+		!GVAR(QTERunning)
+	}, {
+		_this call FUNC(runQTE);
+	}, _this] call CBA_fnc_waitUntilAndExecute;
+};
+
+private _display = findDisplay 46;
+
+if (isNull _display) exitWith {
+	TRACE_1("Waiting for main display to be ready", isNull (_display));
+	[{
+		!isNull (findDisplay 46)
+	}, {
+		_this call FUNC(runQTE);
+	}, _this] call CBA_fnc_waitUntilAndExecute;
+};
+
+if (_onDisplay isEqualType "") then {
+	[_onDisplay, [_args, [_qte_seqence, [_qte_seqence] call FUNC(getFormattedQTESequence)], [[], ""]]] call CBA_fnc_localEvent;
+} else {
+	[_args, [_qte_seqence, [_qte_seqence] call FUNC(getFormattedQTESequence)], [[], ""]] call _onDisplay;
+};
+
+GVAR(QTEHistory) = [];
+GVAR(QTERunning) = true;
+private _start_time = CBA_missionTime;
+
+[_display, "KeyUp", {
+	params ["_displayOrControl", "_key", "_shift", "_ctrl", "_alt"];
+	_thisArgs params ["_object", "_args", "_onDisplay", "_onFinish", "_onFail", "_max_distance", "_qte_seqence", "_start_time", "_timeout", "_display"];
+	private _elapsedTime = CBA_missionTime - _start_time;
+
+	if (player distance _object > _max_distance || _elapsedTime > _timeout) exitWith {
+		_display displayRemoveEventHandler ["KeyUp", _thisID];
+		GVAR(QTERunning) = false;
+		if (_onFail isEqualType "") then {
+			[_onFail, [_args, _elapsedTime]] call CBA_fnc_localEvent;
+		} else {
+			[_args, _elapsedTime] call _onFail;
+		};
+	};
+
+	private _inputKeys = [DIK_UP, DIK_DOWN, DIK_LEFT, DIK_RIGHT];
+	switch (GVAR(QTEInputKeys)) do {
+		case 1: {
+			_inputKeys = [DIK_W, DIK_S, DIK_A, DIK_D];
+		};
+		case 2: {
+			_inputKeys = [DIK_I, DIK_K, DIK_J, DIK_L];
+		};
+		case 3: {
+			_inputKeys = [DIK_NUMPAD8, DIK_NUMPAD2, DIK_NUMPAD4, DIK_NUMPAD6];
+		};
+		default {
+			_inputKeys = [DIK_UP, DIK_DOWN, DIK_LEFT, DIK_RIGHT];
+		};
+	};
```

Comment:
I really dislike this strongly. It uses CBA settings for keybind sets, when we can use CBA keybinds instead.
This would allow for way more flexibility and customisability, as 3rd party mods or mod makers could add their own extra keybinds.

Register 4 keybinds. Here's an example (strings would need to be localised and improved):
```sqf
["QTE Keybinds", QGVAR(keyUpQTE), ["QTE up key", "Up key used in QTE events."], {
    ["↑"] call FUNC(keyPressedQTE); // return
}, {}, [DIK_UP, [false, false, false]]] call CBA_fnc_addKeybind;
```

In `fnc_keyPressedQTE`, you'd check if the keypress is good or not:
```sqf
params ["_eventQTE"];

// Check if the passed parameter is the next key in the QTE sequence
// If yes, continue
// If no, stop
```

Imo, we should make the QTE agonistic of the QTE chars themselves ("↑"). That should be configurable by the user. It will add an extra layer of complexity to the whole things, so we'll have to see if it's worth it.

Reply:
Let's keep it simple and start with assigning keys to "↑" etc then we can look at expanding to other keys.